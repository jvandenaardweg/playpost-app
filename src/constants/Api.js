export const PRODUCTION_API_URL = 'http://medium-audio.herokuapp.com'
export const DEVELOPMENT_API_URL = 'http://localhost:3000'
