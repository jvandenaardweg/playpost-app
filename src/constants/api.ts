const PRODUCTION_API_URL = 'https://api.postplay.app';
const STAGING_API_URL = 'https://api-staging.postplay.app';
const DEVELOPMENT_API_URL = 'http://localhost:3000';

interface ApiEnvironments {
  [key: string]: string;
}

const environments: ApiEnvironments = {
  production: PRODUCTION_API_URL,
  staging: STAGING_API_URL,
  development: DEVELOPMENT_API_URL
};

export const API_URL = environments[process.env.NODE_ENV || 'production'];
