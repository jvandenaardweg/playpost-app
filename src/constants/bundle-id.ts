export const APP_BUNDLE_ID = 'com.aardwegmedia.playpost';
