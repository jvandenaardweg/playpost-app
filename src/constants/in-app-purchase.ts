import { Platform } from 'react-native';

export const SUBSCRIPTION_PRODUCT_ID_FREE = Platform.select({
  ios: 'free',
  android: 'free'
});

export const SUBSCRIPTION_PRODUCT_ID_PREMIUM = Platform.select({
  ios: 'com.aardwegmedia.playpost.premium',
  android: 'com.aardwegmedia.playpost.android.premium'
});

export const SUBSCRIPTION_PRODUCT_ID_PLUS = Platform.select({
  ios: 'com.aardwegmedia.playpost.subscription.plus',
  android: 'com.aardwegmedia.playpost.android.plus'
});

export const SUBSCRIPTION_PRODUCT_IDS = [SUBSCRIPTION_PRODUCT_ID_PREMIUM, SUBSCRIPTION_PRODUCT_ID_PLUS]
