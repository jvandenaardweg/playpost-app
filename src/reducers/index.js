import { articlesReducer } from './articles';
import { playerReducer } from './player';
import { authReducer } from './auth';
import { usersReducer } from './users';

const rootReducer = {
  articles: articlesReducer,
  player: playerReducer,
  auth: authReducer,
  users: usersReducer
};

export default rootReducer;
