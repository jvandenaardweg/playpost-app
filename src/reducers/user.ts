import DeviceInfo from 'react-native-device-info';

import {
  CREATE_USER_FAIL_MESSAGE,
  DELETE_USER_FAIL_MESSAGE,
  GENERIC_NETWORK_ERROR,
  GET_USER_FAIL_MESSAGE,
  PATCH_USER_FAIL_MESSAGE,
  SAVE_SELECTED_VOICE_FAIL_MESSAGE
} from '../constants/messages';

export const GET_USER = 'user/GET_USER';
export const GET_USER_SUCCESS = 'user/GET_USER_SUCCESS';
export const GET_USER_FAIL = 'user/GET_USER_FAIL';

export const PATCH_USER = 'user/PATCH_USER';
export const PATCH_USER_SUCCESS = 'user/PATCH_USER_SUCCESS';
export const PATCH_USER_FAIL = 'user/PATCH_USER_FAIL';

export const CREATE_USER = 'user/CREATE_USER';
export const CREATE_USER_SUCCESS = 'user/CREATE_USER_SUCCESS';
export const CREATE_USER_FAIL = 'user/CREATE_USER_FAIL';

export const DELETE_USER = 'user/DELETE_USER';
export const DELETE_USER_SUCCESS = 'user/DELETE_USER_SUCCESS';
export const DELETE_USER_FAIL = 'user/DELETE_USER_FAIL';

export const SAVE_SELECTED_VOICE = 'user/SAVE_SELECTED_VOICE';
export const SAVE_SELECTED_VOICE_SUCCESS = 'user/SAVE_SELECTED_VOICE_SUCCESS';
export const SAVE_SELECTED_VOICE_FAIL = 'user/SAVE_SELECTED_VOICE_FAIL';
export const RESET_SAVE_SELECTED_VOICE_ERROR = 'user/RESET_SAVE_SELECTED_VOICE_ERROR';


export const SET_PLAYBACK_SPEED = 'user/SET_PLAYBACK_SPEED';

export const RESET_USER_STATE = 'user/RESET_USER_STATE';
export const RESET_USER_ERROR = 'user/RESET_USER_ERROR';

export type UserState = Readonly<{
  isLoading: boolean;
  isLoadingDelete: boolean;
  isLoadingPatchUser: boolean;
  isLoadingSaveSelectedVoice: boolean;
  details: Api.User | null;
  error: string;
  errorSaveSelectedVoice: string;
  deviceLocale: string;
  playbackSpeed: number;
}>;

export const initialState: UserState = {
  isLoading: false,
  isLoadingDelete: false,
  isLoadingPatchUser: false,
  isLoadingSaveSelectedVoice: false,
  details: null,
  error: '',
  errorSaveSelectedVoice: '',
  deviceLocale: DeviceInfo.getDeviceLocale(),
  playbackSpeed: 1
};

/* tslint:disable no-any */
export function userReducer(state = initialState, action: any): UserState {
  switch (action.type) {
    case GET_USER:
      return {
        ...state,
        isLoading: true,
        error: ''
      };

    case GET_USER_SUCCESS:
      return {
        ...state,
        isLoading: false,
        details: action.payload.data,
        error: ''
      };

    case GET_USER_FAIL:
      let getUserFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        getUserFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          getUserFailMessage = action.error.response.data.message;
        } else {
          getUserFailMessage = GET_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoading: false,
        error: getUserFailMessage
      };

    case CREATE_USER:
      return {
        ...state,
        isLoading: true,
        error: ''
      };

    case CREATE_USER_SUCCESS:
      return {
        ...state,
        isLoading: false,
        details: action.payload.data,
        error: ''
      };

    case CREATE_USER_FAIL:
      let createUserFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        createUserFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          createUserFailMessage = action.error.response.data.message;
        } else {
          createUserFailMessage = CREATE_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoading: false,
        details: null,
        error: createUserFailMessage
      };

    case DELETE_USER:
      return {
        ...state,
        isLoadingDelete: true,
        error: ''
      };

    case DELETE_USER_SUCCESS:
      return {
        ...state,
        isLoadingDelete: false,
        error: ''
      };

    case DELETE_USER_FAIL:
      let deleteUserFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        deleteUserFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          deleteUserFailMessage = action.error.response.data.message;
        } else {
          deleteUserFailMessage = DELETE_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoadingDelete: false,
        error: deleteUserFailMessage
      };

    case PATCH_USER:
      return {
        ...state,
        isLoadingPatchUser: true,
        error: ''
      };

    case PATCH_USER_SUCCESS:
      return {
        ...state,
        isLoadingPatchUser: false,
        error: ''
      };

    case PATCH_USER_FAIL:
      let updateUserEmailFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        updateUserEmailFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          updateUserEmailFailMessage = action.error.response.data.message;
        } else {
          updateUserEmailFailMessage = PATCH_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoadingPatchUser: false,
        error: updateUserEmailFailMessage
      };

    case SAVE_SELECTED_VOICE:
      return {
        ...state,
        isLoadingSaveSelectedVoice: true,
        errorSaveSelectedVoice: ''
      };

    case SAVE_SELECTED_VOICE_SUCCESS:
      return {
        ...state,
        isLoadingSaveSelectedVoice: false,
        errorSaveSelectedVoice: ''
      };

    case SAVE_SELECTED_VOICE_FAIL:
      let saveSelectedVoiceFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        saveSelectedVoiceFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          saveSelectedVoiceFailMessage = action.error.response.data.message;
        } else {
          saveSelectedVoiceFailMessage = SAVE_SELECTED_VOICE_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoadingSaveSelectedVoice: false,
        errorSaveSelectedVoice: saveSelectedVoiceFailMessage
      };

    case SET_PLAYBACK_SPEED:
      return {
        ...state,
        playbackSpeed: action.payload
      };

    case RESET_SAVE_SELECTED_VOICE_ERROR:
      return {
        ...state,
        errorSaveSelectedVoice: ''
      };

    case RESET_USER_ERROR:
      return {
        ...state,
        error: ''
      };

    case RESET_USER_STATE:
      return {
        ...initialState
      };

    default:
      return state;
  }
}

export function resetUserState() {
  return {
    type: RESET_USER_STATE
  };
}

export function resetSaveSelectedVoiceError() {
  return {
    type: RESET_SAVE_SELECTED_VOICE_ERROR
  };
}

export function resetUserError() {
  return {
    type: RESET_USER_ERROR
  };
}

export function getUser() {
  return {
    type: GET_USER,
    payload: {
      request: {
        method: 'get',
        url: '/v1/me'
      }
    }
  };
}

export function patchUser({ email, password }: { email?: string, password?: string }) {
  return {
    type: PATCH_USER,
    payload: {
      request: {
        method: 'patch',
        url: '/v1/me',
        data: {
          email,
          password
        }
      }
    }
  };
}

export function createUser(email: string, password: string) {
  return {
    type: CREATE_USER,
    payload: {
      request: {
        method: 'post',
        url: '/v1/users',
        data: {
          email,
          password
        }
      }
    }
  };
}

export function deleteUser() {
  return {
    type: DELETE_USER,
    payload: {
      request: {
        method: 'delete',
        url: '/v1/me'
      }
    }
  };
}

/**
 * Saves the selected voice as a default voice for the voice's language
 *
 * @param voiceId
 */
export function saveSelectedVoice(voiceId: string) {
  return {
    type: SAVE_SELECTED_VOICE,
    payload: {
      request: {
        method: 'post',
        url: '/v1/me/voices',
        data: {
          voiceId
        }
      }
    }
  };
}

export function setPlaybackSpeed(speed: number) {
  return {
    type: SET_PLAYBACK_SPEED,
    payload: speed
  };
}
