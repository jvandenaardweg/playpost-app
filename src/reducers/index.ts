import { combineReducers, Action } from 'redux';
import { AxiosResponse, AxiosRequestConfig } from 'axios';

import { playerReducer } from './player';
import { authReducer } from './auth';
import { userReducer } from './user';
import { playlistReducer } from './playlist';
import { audiofilesReducer } from './audiofiles';
import { voicesReducer } from './voices';

/* tslint:disable no-any */
export interface AxiosActionResponse extends AxiosResponse {
  payload: {
    data: any
  };
}

export interface AxiosAction extends Action {
  payload: {
    request: AxiosRequestConfig
  };
}

const rootReducer = combineReducers({
  player: playerReducer,
  auth: authReducer,
  user: userReducer,
  playlist: playlistReducer,
  audiofiles: audiofilesReducer,
  voices: voicesReducer
});

export type RootState = ReturnType<typeof rootReducer>;

export { rootReducer };
