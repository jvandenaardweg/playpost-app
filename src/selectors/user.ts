import { Platform } from 'react-native';
import { createSelector } from 'reselect';
import { SUBSCRIPTION_NAME, SUBSCRIPTION_PRODUCT_ID_FREE } from '../constants/in-app-purchase';
import { RootState } from '../reducers';
import { UserState } from '../reducers/user';
import { createDeepEqualSelector } from './index';

export interface UserSelectedVoiceByLanguageName {
  [key: string]: Api.Voice
}

export const userSelector = (state: RootState): UserState => state.user;

export const selectUserError = createSelector(
  [userSelector],
  user => user.error
);

export const selectUserErrorSaveSelectedVoice = createSelector(
  [userSelector],
  user => user.errorSaveSelectedVoice
);

export const selectUserIsLoading = createSelector(
  [userSelector],
  user => user.isLoading
);

export const selectUserDetails = createSelector(
  [userSelector],
  user => user.details
);

export const selectUserUsedInAppSubscriptionTrials = createSelector(
  [selectUserDetails],
  userDetails => userDetails && userDetails.usedInAppSubscriptionTrials
);

export const selectDeviceLocale = createSelector(
  [userSelector],
  user => {
    if (!user.deviceLocale) {
      return '';
    }

    // device locale could be: "en-NL" or "en"
    // So we always split the dash
    // Even if the dash does not exist, we return the first in the array
    const deviceLocaleSplitted = user.deviceLocale.split('-');

    if (!deviceLocaleSplitted || !deviceLocaleSplitted[0]) {
      return '';
    }

    return deviceLocaleSplitted[0];
  }
);

export const selectUserSelectedVoices = createDeepEqualSelector(
  [selectUserDetails],
  (userDetails) => {
    if (!userDetails || !userDetails.voiceSettings || !userDetails.voiceSettings.length) {
      return [];
    }
    return userDetails.voiceSettings.map(userVoiceSetting => userVoiceSetting.voice);
  }
);

export const selectUserSelectedVoiceByLanguageName = createDeepEqualSelector(
  [selectUserSelectedVoices],
  (voices): UserSelectedVoiceByLanguageName | null => {
    // Convert the array to an object
    // So we can easily pick a language inside our components
    const voicesObjectWithLanguageNameKeys: UserSelectedVoiceByLanguageName = voices.reduce((prev, curr) => {
      prev[curr.language.name] = curr;

      return prev;
    }, {})

    if (!Object.keys(voicesObjectWithLanguageNameKeys).length) {
      return null;
    }

    return voicesObjectWithLanguageNameKeys;
  }
);

export const selectUserSubscriptions = createDeepEqualSelector(
  [selectUserDetails],
  userDetails => userDetails && userDetails.inAppSubscriptions
);

export const selectUserIsSubscribed = createDeepEqualSelector(
  [selectUserDetails],
  (userDetails): boolean => !!(userDetails && userDetails.activeUserInAppSubscription)
);

/**
 * The active subscription's productId is different per platform.
 *
 */
export const selectUserActiveSubscriptionProductId = createDeepEqualSelector(
  [selectUserDetails, selectUserIsSubscribed],
  (userDetails, isSubscribed): string => {
    if (!isSubscribed) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    if (!userDetails) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    if (!userDetails.activeUserInAppSubscription) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    if (!userDetails.activeUserInAppSubscription.inAppSubscription) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    return userDetails.activeUserInAppSubscription.inAppSubscription.productId
  }
);

export const selectActiveUserInAppSubscription = createDeepEqualSelector(
  [selectUserDetails, selectUserIsSubscribed],
  (userDetails, isSubscribed): Api.UserInAppSubscriptionApple | Api.UserInAppSubscriptionGoogle | null => {
    if (!isSubscribed) { return null; }
    if (!userDetails) { return null; }
    if (!userDetails.activeUserInAppSubscription) { return null; }

    return userDetails.activeUserInAppSubscription;
  }
);

export const selectUserActiveSubscriptionName = createDeepEqualSelector(
  [selectActiveUserInAppSubscription],
  (userActiveSubscription): string => {
    if (!userActiveSubscription) {
      return SUBSCRIPTION_NAME[SUBSCRIPTION_PRODUCT_ID_FREE];
    }
    return userActiveSubscription.inAppSubscription.name
  }
);

export const selectUserIsEligibleForTrial = createDeepEqualSelector(
  [selectUserUsedInAppSubscriptionTrials],
  (usedInAppSubscriptionTrials): boolean => {
    if (!usedInAppSubscriptionTrials || !usedInAppSubscriptionTrials.length) {
      return true;
    }

    // If the platform is Android, look up if the user had a Google subscription trial before
    if (Platform.OS === 'android') {
      return !!!usedInAppSubscriptionTrials.find(usedInAppSubscription => usedInAppSubscription.service === 'google')
    }

    // If the platform is else (iOS), look up if the user had a Apple subscription trial before
    return !!!usedInAppSubscriptionTrials.find(usedInAppSubscription => usedInAppSubscription.service === 'apple')
  }
);

export const selectUserHasSubscribedBefore = createDeepEqualSelector(
  [selectUserSubscriptions],
  userInAppSubscriptions => {
    if (!userInAppSubscriptions || !userInAppSubscriptions.length) {
      return false;
    }

    // If the user has a previous in app subscription record, return true
    return !!userInAppSubscriptions.length;
  }
);

export const selectUserPlaybackSpeed = createSelector(
  [userSelector],
  user =>  user.playbackSpeed
)
