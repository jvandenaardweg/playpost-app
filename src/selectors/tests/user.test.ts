import { createStore } from 'redux';

import { selectDeviceLocale,
  selectUserActiveSubscriptionName,
  selectUserActiveSubscriptionProductId,
  selectUserDetails,
  selectUserError,
  selectUserErrorSaveSelectedVoice,
  selectUserHasSubscribedBefore,
  selectUserIsLoading,
  selectUserIsSubscribed,
  selectUserPlaybackSpeed,
  selectUserSelectedVoiceByLanguageName,
  selectUserSelectedVoices,
  selectUserSubscriptions,
  userSelector
} from '../user';

import { rootReducer } from '../../reducers';
import { initialState } from '../../reducers/user';

import exampleUserWithActiveSubscription from '../../../tests/__mocks__/user-active-subscription';
import exampleUserWithInactiveSubscription from '../../../tests/__mocks__/user-inactive-subscription';
import { SUBSCRIPTION_PRODUCT_ID_FREE, SUBSCRIPTION_PRODUCT_ID_PREMIUM_ANDROID } from '../../constants/in-app-purchase';

const store = createStore(rootReducer);

const rootState = store.getState();
// const userStore = rootState.user;

describe('user selector', () => {
  it('should return the initial state', () => {
    expect(userSelector(rootState)).toEqual(initialState);
  });

  it('should return the user error', () => {
    const exampleErrorState = {
      ...rootState,
      user: {
        ...rootState.user,
        error: 'Test error'
      }
    };

    expect(selectUserError(exampleErrorState)).toBe('Test error');
  });

  it('selectUserErrorSaveSelectedVoice should return the user error', () => {
    const exampleErrorState = {
      ...rootState,
      user: {
        ...rootState.user,
        errorSaveSelectedVoice: 'Test error 2'
      }
    };

    expect(selectUserErrorSaveSelectedVoice(exampleErrorState)).toBe('Test error 2');
  });

  it('should return the loading status', () => {
    const exampleLoadingState = {
      ...rootState,
      user: {
        ...rootState.user,
        isLoading: true
      }
    };

    expect(selectUserIsLoading(exampleLoadingState)).toBe(true);
  });

  it('should return the user details', () => {

    const exampleUserDetailsState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithActiveSubscription
      }
    };

    expect(selectUserDetails(exampleUserDetailsState)).toMatchObject(exampleUserWithActiveSubscription);
  });

  it('should return the user\'s selected voices', () => {
    const exampleUserDetailsState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithActiveSubscription
      }
    };

    const expected = exampleUserWithActiveSubscription.voiceSettings.map(userVoiceSetting => userVoiceSetting.voice);

    expect(selectUserSelectedVoices(exampleUserDetailsState)).toEqual(expected);

    const exampleUserDetailsState2 = {
      ...rootState,
      user: {
        ...initialState
      }
    };
    expect(selectUserSelectedVoices(exampleUserDetailsState2)).toEqual([]);
  });

  it('should return the user\'s selected voices by language name', () => {
    const exampleUserDetailsState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithActiveSubscription
      }
    };

    expect(selectUserSelectedVoiceByLanguageName(exampleUserDetailsState)).toMatchSnapshot();
  });

  it('should return the user\'s subscriptions', () => {
    const exampleUserState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithActiveSubscription
      }
    };

    const expected = exampleUserWithActiveSubscription.inAppSubscriptions;

    expect(selectUserSubscriptions(exampleUserState)).toEqual(expected);
  });

  it('selectDeviceLocale should return the user\'s subscriptions', () => {
    const exampleUserStateOne = {
      ...rootState,
      user: {
        ...rootState.user,
        deviceLocale: 'en-NL'
      }
    };

    const exampleUserStateTwo = {
      ...rootState,
      user: {
        ...rootState.user,
        deviceLocale: 'en'
      }
    };

    const exampleUserStateThree = {
      ...rootState,
      user: {
        ...rootState.user,
        deviceLocale: ''
      }
    };

    expect(selectDeviceLocale(exampleUserStateOne)).toEqual('en');
    expect(selectDeviceLocale(exampleUserStateTwo)).toEqual('en');
    expect(selectDeviceLocale(exampleUserStateThree)).toEqual('');
  });

  it('selectUserHasSubscribedBefore should return true when the user has previous subscriptions', () => {
    const exampleUserState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithActiveSubscription
      }
    };

    expect(selectUserHasSubscribedBefore(exampleUserState)).toEqual(true);
  });

  it('selectUserHasSubscribedBefore should return false when the user has no previous subscriptions', () => {
    const exampleUserCopy = {...exampleUserWithActiveSubscription};

    exampleUserCopy.inAppSubscriptions = [];

    const exampleUserStateWithoutSubscriptions = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserCopy
      }
    };

    expect(selectUserHasSubscribedBefore(exampleUserStateWithoutSubscriptions)).toEqual(false);
  });

  it('selectUserPlaybackSpeed should return a number', () => {
    const exampleUserState = {
      ...rootState,
      user: {
        ...rootState.user,
        playbackSpeed: 0.95
      }
    };

    expect(selectUserPlaybackSpeed(exampleUserState)).toEqual(0.95);
  });

  it('selectUserIsSubscribed should return false when there is no user details', () => {
    const exampleState = {
      ...rootState,
      user: {
        ...rootState.user
      }
    };

    expect(selectUserIsSubscribed(exampleState)).toEqual(false);
  });

  it('selectUserIsSubscribed should return true when there is an active subscription', () => {
    const exampleState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithActiveSubscription
      }
    };

    expect(selectUserIsSubscribed(exampleState)).toEqual(true);
  });

  it('selectUserIsSubscribed should return false when there is no active subscription', () => {
    const exampleState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithInactiveSubscription
      }
    };

    expect(selectUserIsSubscribed(exampleState)).toEqual(false);
  });


  it('selectUserActiveSubscriptionProductId should return the default productId "free" when there are no active subscriptions', () => {
    const exampleState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithInactiveSubscription
      }
    };

    // The mock data contains a unsubscribed/expired subscription
    expect(selectUserActiveSubscriptionProductId(exampleState)).toEqual(SUBSCRIPTION_PRODUCT_ID_FREE);
  });

  it('selectUserActiveSubscriptionProductId should return the active subscription productId', () => {
    const exampleState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithActiveSubscription
      }
    };

    // The mock data contains a unsubscribed/expired subscription
    expect(selectUserActiveSubscriptionProductId(exampleState)).toEqual(SUBSCRIPTION_PRODUCT_ID_PREMIUM_ANDROID);
  });

  it('selectUserActiveSubscriptionProductId should return the default productId "free" when the user has no active subscription', () => {
    const exampleState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithInactiveSubscription
      }
    };

    // The mock data contains a unsubscribed/expired subscription
    expect(selectUserActiveSubscriptionProductId(exampleState)).toEqual(SUBSCRIPTION_PRODUCT_ID_FREE);
  });

  it('selectUserActiveSubscriptionProductId should return the default productId "free" when there is no user info', () => {
    const exampleState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: null
      }
    };

    // The mock data contains a unsubscribed/expired subscription
    expect(selectUserActiveSubscriptionProductId(exampleState)).toEqual(SUBSCRIPTION_PRODUCT_ID_FREE);
  });

  it('selectUserActiveSubscriptionName should return the active subscription name when the user has a validated subscription', () => {
    const exampleState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: exampleUserWithActiveSubscription
      }
    };

    // The mock data contains a unsubscribed/expired subscription
    expect(selectUserActiveSubscriptionName(exampleState)).toEqual('Premium');
  });

  it('selectUserActiveSubscriptionName should return "Free" if the user has no subscription', () => {
    const exampleState = {
      ...rootState,
      user: {
        ...rootState.user,
        details: null
      }
    };

    // The mock data contains a unsubscribed/expired subscription
    expect(selectUserActiveSubscriptionName(exampleState)).toEqual('Free');
  });


});
