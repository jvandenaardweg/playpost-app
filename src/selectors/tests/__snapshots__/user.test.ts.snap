// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user selector should return the user's selected voices by language name 1`] = `null`;
